addons/udes_stock/models/stock_picking.py

```diff
@@ -2783,130 +2784,137 @@ def raise_insufficiency_error(moves):
+                    pickings_to_process = self.filtered(by_type) - processed
                 else:
                     domain = base_domain[:]
                     if processed:
                         domain.append(("id", "not in", processed.ids))
-                    pickings = Picking.search(domain, limit=limit)
+                    pickings_to_process = Picking.search(domain, limit=limit)
 
-                if not pickings:
+                if not pickings_to_process:
                     # No pickings left to process.
                     # If u_num_reservable_pickings is -1, or there are
                     # fewer available pickings that the limit, the loop must
                     # terminate here.
                     break
 
-                batch = pickings.mapped("batch_id")
-                if batch and batch.state == "draft":
-                    # Add to seen pickings so that we don't try to process
-                    # this batch again.
-                    processed |= batch.picking_ids
-                    continue
-
-                if batch and picking_type.u_reserve_batches:
-                    pickings = batch.picking_ids
-
-                # Check for available stock and add pickings which we cannot
-                # fulfill to the processed recordset. The logic here
-                # should mirror that within the savepoint below, expect here we
-                # do not have the move lines.
-                unreservable_moves = pickings._find_unreservable_moves(return_fast=bool(self))
-                if unreservable_moves:
-                    moves = pickings.mapped("move_lines")
-                    moves_todo = (
-                        moves.filtered(lambda m: m.state in ["confirmed", "draft"])
-                        - unreservable_moves
-                    )
-                    if not picking_type.u_handle_partials:
-                        # Add to processed to skip reservation.
-                        processed |= pickings
-
-                        if self:
-                            raise_insufficiency_error(unreservable_moves)
-                        continue
-                    elif not moves_todo:
-                        # Add to processed to skip reservation.
-                        processed |= pickings
+                # "pickings" is plural as the individual picking may be
+                # extended if it's in a batch.
+                for pickings in pickings_to_process:
+                    if pickings in processed:
                         continue
 
-                # MPS: mimic Odoo's retry behaviour
-                tries = 0
-                while True:
-
-                    try:
-                        with self.env.cr.savepoint():
-                            pickings = pickings._reserve_stock_assign()
-                            batch._compute_state()
+                    batch = pickings.mapped("batch_id")
```

Comment:
It seems like now we are doing batch by batch, maybe we should group by the picks who have batch and who hasnt.

Reply:
I just indented the existing code.  Fixed by 1f0282b2a.